graph.cpp

```diff
@@ -141,6 +141,89 @@ vector<set<Node>> Graph::connected_components() {
   return retval;
 }
 
+vector<set<Node>> Graph::parallel_connected_components() {
+#ifdef USE_FBT_F
+  bf->force_flush(); // flush everything in buffertree to make final updates
+#else
+  wq->force_flush();
+#endif
+  GraphWorker::pause_workers(); // wait for the workers to finish applying the updates
+  // after this point all updates have been processed from the buffer tree
+
+  printf("Total number of updates to sketches before CC %lu\n", num_updates.load()); // REMOVE this later
+  update_locked = true; // disallow updating the graph after we run the alg
+  bool modified;
+#ifdef VERIFY_SAMPLES_F
+  GraphVerifier verifier { cum_in };
+#endif
+  pair<Node,Node> query[num_nodes];
+  Node size[num_nodes];
+  vector<Node> reps(num_nodes);
+  fill(size, size + num_nodes, 1);
+  for (Node i = 0; i < num_nodes; ++i) {
+    reps[i] = i;
+  }
+
+  do {
+    modified = false;
+    #pragma omp parallel for default(none) shared(query, reps)
+    for (Node i = 0; i < reps.size(); ++i) {
+      auto edge = supernodes[reps[i]]->sample();
+      if (!edge.is_initialized()) {
+        query[reps[i]] = {i,i};
+        continue;
+      }
+      query[reps[i]] = edge.get();
+    }
+
+    vector<Node> to_remove;
+    for (Node i : reps) {
+      Node a = get_parent(query[i].first);
+      Node b = get_parent(query[i].second);
+      if (a == b) continue;
+#ifdef VERIFY_SAMPLES_F
+      verifier.verify_edge({query[i].first,query[i].second});
+#endif
+      // make sure a is the one to be merged into
+      if (size[a] < size[b]) std::swap(a,b);
+      to_remove.push_back(b);
+      parent[b] = a;
+      size[a] += size[b];
+      supernodes[a]->merge(*supernodes[b]);
+    }
+    if (!to_remove.empty()) modified = true;
+    sort(to_remove.begin(), to_remove.end());
+
+    // 2-pointer to find set difference
```

Comment:
Can you explain what this set difference thing is doing?

Reply:
Essentially, we're treating `reps` and `to_remove` as sets, and we're finding `temp_diff = reps \ to_remove`

Reply:
Gotcha